### ISSUE TITLE
Remove context header for anonymous users

### ISSUE THREAD

**Issue by ddeboer** [FriendsOfSymfony/FOSHttpCache#182]
From #180:

> The hash request is only ever executed when there is a cookie / auth header: https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/tests/Functional/Fixtures/varnish-3/user_context.vcl#L16
> 
> The Vary information also includes the variant of not having the header in question.
> 
> I think we can close #132 and think whether we should remove the corresponding code in the SymfonyCache and just have no context header. https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/src/SymfonyCache/UserContextSubscriber.php#L165 - the backend certainly should not use the hash, but only Vary on it.

**Comment by ddeboer** [comment]
Sounds like a good idea to me!

@lolautruche Do you agree (based on your comments in #132)?

**Comment by dbu** [comment]
would make sense to me.

**Comment by dbu** [comment]
@lolautruche wdyt? 

the alternative scenario would be to actually go to the backend even without cookie (only once, then that hash is cached too) in case there might be a session for non-logged-in users or even logged in users that belong to the same group as anonymous users. with the manual hash, we split this group.

**Comment by ddeboer** [comment]
We should remember that Symfony/Silex currently automatically creates a session, and thus a PHPSESSID cookie. This means that even for anonymous users a Cookie header will be present and (according to our Varnish config) the hash lookup will take place.

@dbu There’s been some discussion about this, and if I’m not mistaken you were part of that. Do you know which decisions have been reached concerning Symfony’s behaviour?

**Comment by dbu** [comment]
There is a cookbook entry on how to avoid session in symfony. I think the check would just be a shortcut for when no session for anonymous. We can document it like that - if you do session for anonymous, its just useless overhead.

I think it will never create a problem - unless you ignore but use basic auth, or some custom authentication system without cookies...

**Comment by lolautruche** [comment]
@ddeboer Symfony doesn't automatically create a session. It does when using `dev` environment (profiler).

> the alternative scenario would be to actually go to the backend even without cookie (only once, then that hash is cached too) in case there might be a session for non-logged-in users or even logged in users that belong to the same group as anonymous users.

@dbu If there is a session (even anonymous), we *always* go to the backend. This is the only possible discrimination. As for authenticated users who belong to the same group as anonymous users, we don't split them. It's just that they will go to the backend for them as they do have a session.

TBH, I don't see any benefit of removing the hash for anonymous users.

**Comment by dbu** [comment]
ok i was confusing. my point was that maybe the backend could decide on giving the same hash when there is no session as when there is session for anonymous as when i am a authenticated but limited user.

i think the user has two options: 
- prevent going to the backend when there is no session. whether we hardcode a header or not set a header at all makes technically no difference. my suggestion was to remove the header in this case as it has no meaning.
- still go to the backend for hash lookup. this should be only one request which is then cached, so the performance impact should be extremly small (unless hash lookups are not cached)

given that, i actually think that we should by default always go to the backend. it removes one potential WTF at the cost of 1 request per hour or less.

**Comment by dbu** [comment]
btw, regarding sessions in symfony, there is http://symfony.com/doc/current/cookbook/session/avoid_session_start.html

**Comment by dbu** [comment]
> given that, i actually think that we should by default always go to the backend for the hash lookup. 
> it removes one potential WTF at the cost of 1 request per hour or less.

@ddeboer @lolautruche as we are talking about tagging 1.4 - what about this one? would be nice to decide before we release.

at the same time we can explain how to avoid the call to the backend with varnish if not needed. that is just documentation, no actual code as its all in vcl.

**Comment by ddeboer** [comment]
I decided not to wait with 1.4 until this is solved.

**Comment by dbu** [comment]
agreed. i will create a version 2 milestone and add this issue to it as we should clean things up for 2.0 (there is some wtf potential here - both with the current behaviour but also of course when changing default behaviour)

**Comment by dbu** [comment]
@lolautruche we are slowly thinking of wrapping up a 2.0 of this library. what do you think of this? i think the cleanest would be to always do a request for the hash, even when its an anonymous user. the hash requests are cached so it costs very little and avoids some confusion. does that make sense to you?

whichever we go, i would love to make varnish and httpcache consistent in this.

**Comment by dbu** [comment]
ping @lolautruche - unless you convince me otherwise, i will go for always doing the request, also with httpcache. as said above, its a tiny overhead (cached after the first request) that can even lead to save requests (anonymous vs non-privileged session can share the content cache) and avoids a wtf moment for devs by removing a special case on the caching proxy.

**Comment by lolautruche** [comment]
Hi

Sorry for not replying before.
Actually, the hardcoded hash was here for performance reasons with eZ, as generating the hash has a cost since eZ kernel must load and loop over user permissions. So this greatly speeded up things for us.
However, I understand what can be claimed here. eZ kernel can still use this header for itself or avoid generating the hash if it detects that the user is anonymous during the hash request.

Either way, I'm fine with it as soon as it is documented as BC break.

Ping @bdunogier @andrerom

**Comment by andrerom** [comment]
As long as the hash is always cached, we're probably ok with changing this for 2.0.

Why:
- we had (https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/pull/289), and still have issues when customers change their anonymous users and it will thus have thus mean anonymous and anonymous with session cookie does not share cache


So by making sure we:
- always cache the response _(btw people at Varnish suggests we rather use curl to varnish itself for caching hash lookup, instead of restart, but that's another story for another thread)_
- tag the hash response so we can reliably clear it on relevant changes, and thus cache it for much longer time

Then I _think_ we are ok.
But @bdunogier has been more deeply digging into this then me at times so pinging him on this to.

**Comment by ddeboer** [comment]
> btw people at Varnish suggests we rather use curl to varnish itself for caching hash lookup, instead of restart, but that's another story for another thread)

Interesting, care to open up a separate issue for that and elaborate a bit?

**Comment by andrerom** [comment]
@lolautruche As you spoke more with him, are you able to verbalize the issue he raised?

**Comment by dbu** [comment]
>     btw people at Varnish suggests we rather use curl to varnish itself
>     for caching hash lookup, instead of restart, but that's another
>     story for another thread)
> 
> Interesting, care to open up a separate issue for that and elaborate a bit?

there is a curl vmod that could be used to do the hash look without
restarting the request. i agree that would be better.

the reason we did not do that is that it would require the curl vmod to
be installed. its not installed by default. imho we would need to offer
a separate version of the vcl for when you have the curl vmod.

**Comment by ddeboer** [comment]
Correct. If I’m not mistaken, this would be mostly a doc/test issue as we shouldn’t have to change our implementation code.

**Comment by dbu** [comment]
or that, yes. we could provide a tested vcl for the use case though. if we get varnish set up with the curl vmod on travis-ci

**Comment by ddeboer** [comment]
> if we get varnish set up with the curl vmod on travis-ci

That would require building Varnish from source.

**Comment by dbu** [comment]
ok, then rather not :-) happy if somebody wants to do a PR to add the curl variant to the documentation.
 
back to this PR: 
> tag the hash response so we can reliably clear the hash lookup itself, and not the whole cache on permission changes, and thus also cache it for much longer time reducing the cost of the lookup as long as no one changes permissions all the time.

should we copy the cookie header onto the cache object to enable banning? (you could ban on requests having no cookie, but that would not play well with the ban lurker)

**Comment by dbu** [comment]
sorry, now i was confused. we should continue the discussion in #318

**Comment by ddeboer** [comment]
This was fixed in #330.
